cirq-core/cirq/optimizers/transformer_primitives.py

```diff
@@ -0,0 +1,186 @@
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Moment, moment_index) to a sequence of moments.
+    """
+    return _to_target_circuit_type(
+        circuits.Circuit(map_func(circuit[i], i) for i in range(len(circuit))), circuit
+    )
+
+
+def map_operations(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Operation, int], ops.OP_TREE],
+) -> CIRCUIT_TYPE:
+    """Applies local transformations on operations, by calling `map_func(op)` for each op.
+
+    Note that the function assumes `issubset(qubit_set(map_func(op)), op.qubits)` is True.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Operation, moment_index) to a cirq.OP_TREE. If the
+            resulting optree spans more than 1 moment, it's inserted in-place in the same moment as
+            `cirq.CircuitOperation(cirq.FrozenCircuit(op_tree)).with_tags(MAPPED_CIRCUIT_OP_TAG)`
+            to preserve moment structure. Utility methods like `cirq.unroll_circuit_op` can
+            subsequently be used to unroll the mapped circuit operation.
+
+    Raises:
+          ValueError if `issubset(qubit_set(map_func(op)), op.qubits) is False`.
+    """
+
+    def apply_map(op: ops.Operation, idx: int) -> ops.OP_TREE:
+        c = circuits.FrozenCircuit(map_func(op, idx))
+        if not c.all_qubits().issubset(op.qubits):
+            raise ValueError(
+                f"Mapped operations {c.all_operations()} should act on a subset "
+                f"of qubits of the original operation {op}"
+            )
+        if len(c) == 1:
+            # All operations act in the same moment; so we don't need to wrap them in a circuit_op.
+            return c[0].operations
+        circuit_op = circuits.CircuitOperation(c).with_tags(MAPPED_CIRCUIT_OP_TAG)
+        return circuit_op
+
+    return map_moments(circuit, lambda m, i: [ops.Moment(apply_map(op, i) for op in m.operations)])
+
+
+def map_operations_and_unroll(
+    circuit: CIRCUIT_TYPE,
+    map_func: Callable[[ops.Operation, int], ops.OP_TREE],
+) -> CIRCUIT_TYPE:
+    """Applies local transformations via `cirq.map_operations` & unrolls intermediate circuit ops.
+
+    See `cirq.map_operations` and `cirq.unroll_circuit_op` for more details.
+
+    Args:
+        circuit: Input circuit to apply the transformations on. The input circuit is not mutated.
+        map_func: Mapping function from (cirq.Operation, moment_index) to a cirq.OP_TREE.
+    """
+    return unroll_circuit_op(map_operations(circuit, map_func))
+
+
+def _check_circuit_op(op, tags_to_check: Optional[Sequence[Hashable]]):
+    return isinstance(op.untagged, circuits.CircuitOperation) and (
+        tags_to_check is None or any(tag in op.tags for tag in tags_to_check)
+    )
+
+
+def unroll_circuit_op(
```

Comment:
Can we merge these three unroll_circuit_op_*** into one ?

Reply:
I don't think there's a clean way to merge the `unrol_circuit_op**` into a single method because 2/3 implementations do not correspond to a standard insert strategy and hence the insert strategy parameter would be misleading at best. We can come up with another enum and pass it as a parameter but I don't see the advantage, especially because the implementations of the 3 methods are also very different.